app/src/main/java/nerd/tuxmobil/fahrplan/congress/changes/ChangesDialog.kt

```diff
@@ -0,0 +1,129 @@
+import nerd.tuxmobil.fahrplan.congress.repositories.AppRepository.updateScheduleChangesSeen
+
+import android.app.Activity
+import android.app.Dialog
+import android.content.DialogInterface
+import android.os.Bundle
+import android.text.Spannable
+import android.text.SpannableStringBuilder
+import android.text.style.ForegroundColorSpan
+import android.widget.TextView
+
+import androidx.annotation.MainThread
+import androidx.appcompat.app.AlertDialog
+import androidx.core.content.ContextCompat
+import androidx.fragment.app.DialogFragment
+
+import nerd.tuxmobil.fahrplan.congress.R
+import nerd.tuxmobil.fahrplan.congress.contract.BundleKeys
+import nerd.tuxmobil.fahrplan.congress.schedule.MainActivity
+
+class ChangesDialog : DialogFragment() {
+
+    companion object {
+        const val FRAGMENT_TAG = "changesDialog"
+        fun newInstance(
+            version: String,
+            statistic: ChangeStatistic
+        ): ChangesDialog {
+            val dialog = ChangesDialog()
+            val args = Bundle()
+            args.putInt(BundleKeys.CHANGES_DLG_NUM_CHANGED, statistic.getChangedSessionsCount())
+            args.putInt(BundleKeys.CHANGES_DLG_NUM_NEW, statistic.getNewSessionsCount())
+            args.putInt(BundleKeys.CHANGES_DLG_NUM_CANCELLED, statistic.getCanceledSessionsCount())
+            args.putInt(BundleKeys.CHANGES_DLG_NUM_MARKED, statistic.getChangedFavoritesCount())
+            args.putString(BundleKeys.CHANGES_DLG_VERSION, version)
+            dialog.arguments = args
+            dialog.isCancelable = false
+            return dialog
+        }
+    }
+
+    private var changed = 0
+    private var added = 0
+    private var cancelled = 0
+    private var markedAffected = 0
+    private var version: String = ""
+
+    @MainThread
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val args = requireArguments()
+
+        changed = args.getInt(BundleKeys.CHANGES_DLG_NUM_CHANGED)
+        added = args.getInt(BundleKeys.CHANGES_DLG_NUM_NEW)
+        cancelled = args.getInt(BundleKeys.CHANGES_DLG_NUM_CANCELLED)
+        markedAffected = args.getInt(BundleKeys.CHANGES_DLG_NUM_MARKED)
+        version = args.getString(BundleKeys.CHANGES_DLG_VERSION)!!
+    }
+
+    @MainThread
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        val activity: Activity = requireActivity()
+
+        val builder = AlertDialog.Builder(activity).apply {
+            setTitle(getString(R.string.schedule_changes_dialog_title))
+            setPositiveButton(R.string.schedule_changes_dialog_browse) { _: DialogInterface?, _: Int -> onBrowse() }
+            setNegativeButton(R.string.schedule_changes_dialog_later) { _: DialogInterface?, _: Int -> onLater() }
+        }
+
+        val msgView = activity.layoutInflater.inflate(R.layout.changes_dialog, null)
+
+        val span = SpannableStringBuilder()
```

Comment:
Have a look at `androidx.core.text.*` to improve the "span" code.

Reply:
@isiko Are you still looking into if the `span` construction can be simplified?

Reply:
Jupp, am Looking into that. This should probably be done using the `<annotation>` in the `strings.xml` file. Then we could just load one String and then apply the formatting instead of appending a lot of stuff to one another. [Here](https://developer.android.com/guide/topics/resources/string-resource#StylingWithAnnotations)'s a link to an example from the Documentation.

I'll try to implement that, but as I would have to change some stuff in the `strings.xml` file and don't have that much experience with translation in Android Apps, I'm gonna have to be very careful to not break anything.

Reply:
Also: Is there a way to get that Dialog without any running Events?

Reply:
Did you have a look at the Kotlin extension functions provided here?
- https://github.com/androidx/androidx/blob/androidx-main/core/core-ktx/src/main/java/androidx/core/text/SpannableString.kt
- https://github.com/androidx/androidx/blob/androidx-main/core/core-ktx/src/main/java/androidx/core/text/SpannableStringBuilder.kt
- https://github.com/androidx/androidx/blob/androidx-main/core/core-ktx/src/main/java/androidx/core/text/SpannedString.kt
My idea was that the might help to simplify the Kotlin implementation. I did not try myself.

I never work with `Annotations`. That seems to be a new API. If you want you can give it a try in a follow-up pull request.

The easiest would be to just launch the dialog manually/temporarily in `MainActivity` or so.

Reply:
I looked into that, but I don't see that much room to improve here. Appending and highlighting the Version String could be combined into one line, but that function is only available on API Level 21 or higher, but that's it as far as I see.

Also, as it might be a good idea to reimplement that Code using Annotations (as mentioned above) in the Future, this Code might not be used for long anyway.